lib/teleterm/daemon/daemon.go

```diff
@@ -269,11 +270,53 @@ func (s *Service) SetGatewayTargetSubresourceName(gatewayURI, targetSubresourceN
 		return nil, trace.Wrap(err)
 	}
 
-	gateway.TargetSubresourceName = targetSubresourceName
+	gateway.SetTargetSubresourceName(targetSubresourceName)
 
 	return gateway, nil
 }
 
+// SetGatewayLocalPort creates a new gateway with the given port, swaps it with the old gateway
+// under the same URI in s.gateways and then closes the old gateway. It doesn't fetch a fresh db
+// cert.
+//
+// If it gateway.NewWithLocalPort fails it's imperative that the current gateway is kept intact.
+// This way if the user attempts to change the port to one that cannot be obtained, they're able to
+// correct that mistake and choose a different port.
+//
+// SetGatewayLocalPort is a noop if port is equal to the existing port.
+func (s *Service) SetGatewayLocalPort(gatewayURI, localPort string) (*gateway.Gateway, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	oldGateway, err := s.findGateway(gatewayURI)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if localPort == oldGateway.LocalPort() {
+		return oldGateway, nil
+	}
+
+	newGateway, err := gateway.NewWithLocalPort(*oldGateway, localPort)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := s.removeGateway(oldGateway); err != nil {
+		return nil, trace.Wrap(err)
+	}
```

Comment:
My thinking behind returning an error here was that if closing the old gateway fails, it'll still remain in the hash map so the Electron app will see it as active. The user will see that the attempt to change the port has failed and will be able to retry it.

If we continued to use the new one then the old one would get lost in a limbo? 🤔

Failing to close a gateway boils down to failing to close a listener and to be honest I have no idea why it would fail and what side effects it would have.

---

But! I thought about this scenario and my WIP version even included closing the new gateway here. In the end I either removed it by omission or maybe I wasn't sure what to do with the error from `newGateway.Close()`. Should I just log it and move on?

Reply:
a54c0ea closes the new gateway and logs the error.

Reply:
IDK what's the best action either. If we show the close error to user, and user retires, will that fix closing the old gateway? But like you said, not sure why would closing a listener fail in the first place.

Reply:
> But like you said, not sure why would closing a listener fail in the first place.

I dove into the source code and, at least for TCP listeners, it seems that `Close()` fails if [the listener was already closed](https://cs.opensource.google/go/go/+/refs/tags/go1.18.4:src/internal/poll/fd_mutex.go;drc=2ff563a00e45308fad74f1f86d816f4f636584a0;l=71). Here's a short test which fails because of the second `Close()`:

```go
func TestNetClose(t *testing.T) {
	lis, err := net.Listen("tcp", "localhost:0")
	require.NoError(t, err)
	require.NoError(t, lis.Close())
	require.NoError(t, lis.Close())
}
```

Then there's [this `l.ok()` check](https://cs.opensource.google/go/go/+/refs/tags/go1.18.4:src/net/tcpsock.go;l=298). It's called from almost every `TCPListener` method so I assume this is some internal consistency check or sth.

---

In any case, if this call fails that means we have a race condition and the listener was already closed. In that case, it'd be best to log the error and use the new listener. I'll change the code accordingly.

Reply:
Or actually, since we do have locks in place and `gateway.Close()` should never fail, failing to close the old one would mean we have a race condition. In this scenario, I believe it'd be better to fail loudly rather than let the app continue running despite the race condition.

So I'll leave the behavior unchanged, that is if closing the old gateway fails, close the new one too, log the error from closing the new one if any and then report the error from closing the old one. But I'll update the comments so that they explain why we do this.